classes/migration/upgrade/v3_4_0/I6782_OrphanedMetrics.inc.php

```diff
@@ -0,0 +1,123 @@
+use PKP\migration\Migration;
+
+abstract class I6782_OrphanedMetrics extends Migration
+{
+    abstract protected function getContextTable(): string;
+    abstract protected function getContextKeyField(): string;
+    abstract protected function getRepresentationTable(): string;
+    abstract protected function getRepresentationKeyField(): string;
+
+    /**
+     * Run the migration.
+     *
+     * assoc_object_type, assoc_object_id, and pkp_section_id will not be considered here, because they are not relevant for the migration
+     */
+    public function up(): void
+    {
+        $this->createMetricsTmpTable();
+
+        $metricsColumns = Schema::getColumnListing('metrics_tmp');
+
+        // Clean orphaned metrics context IDs. These IDs can be deleted.
+        // as assoc_id
+        $orphanedIds = DB::table('metrics AS m')->leftJoin($this->getContextTable() . ' AS c', 'm.assoc_id', '=', 'c.' . $this->getContextKeyField())->where('m.assoc_type', '=', Application::getContextAssocType())->whereNull('c.' . $this->getContextKeyField())->distinct()->pluck('m.assoc_id');
+        foreach ($orphanedIds as $contextId) {
+            $this->_installer->log("Removing stats for context {$contextId} because no context with that ID could be found.");
+            DB::table('metrics')->where('assoc_type', '=', Application::getContextAssocType())->where('assoc_id', '=', $contextId)->delete();
+        }
+        // as context_id
+        $orphanedIds = DB::table('metrics AS m')->leftJoin($this->getContextTable() . ' AS c', 'm.context_id', '=', 'c.' . $this->getContextKeyField())->whereNull('c.' . $this->getContextKeyField())->distinct()->pluck('m.context_id');
+        foreach ($orphanedIds as $contextId) {
+            $this->_installer->log("Removing stats for context {$contextId} because no context with that ID could be found.");
+            DB::table('metrics')->where('context_id', '=', $contextId)->delete();
+        }
+
+        // Clean orphaned metrics submission IDs
+        // as submission_id
+        $orphanedIds = DB::table('metrics AS m')->leftJoin('submissions AS s', 'm.submission_id', '=', 's.submission_id')->whereNotNull('m.submission_id')->whereNull('s.submission_id')->distinct()->pluck('m.submission_id');
+        $orphandedSubmissions = DB::table('metrics')->select('*')->whereIn('submission_id', $orphanedIds);
+        DB::table('metrics_tmp')->insertUsing($metricsColumns, $orphandedSubmissions);
+        DB::table('metrics')->whereIn('submission_id', $orphanedIds)->delete();
+
+        // as assoc_id
+        $orphanedIds = DB::table('metrics AS m')->leftJoin('submissions AS s', 'm.assoc_id', '=', 's.submission_id')->where('m.assoc_type', '=', Application::ASSOC_TYPE_SUBMISSION)->whereNull('s.submission_id')->distinct()->pluck('m.assoc_id');
+        $orphandedSubmissionsAssocId = DB::table('metrics')->select('*')->where('assoc_type', '=', Application::ASSOC_TYPE_SUBMISSION)->whereIn('assoc_id', $orphanedIds);
```

Comment:
Lately we've been duplicating constants in the migration classes -- see e.g.:

https://github.com/pkp/pkp-lib/blob/a4d500a74589f2a008dc29c11938d3b7fe0fa1cb/classes/migration/upgrade/PKPv3_2_1SubeditorCategoriesMigration.inc.php#L22

This ensures that the migrations use the values for constants that are appropriate for the version of the software that the upgrade is intended for. If the constant later changes (e.g. if we harmonize some of the `ASSOC_TYPE_...` values across different applications), those constants can change without breaking migrations that might otherwise have relied on them.

Reply:
I also added the protected function getContextAssocType()...